core/src/main/scala/org/apache/spark/Partitioner.scala

```diff
@@ -67,31 +69,32 @@ object Partitioner {
       None
     }
 
-    if (isEligiblePartitioner(hasMaxPartitioner, rdds)) {
+    val defaultNumPartitions = if (rdd.context.conf.contains("spark.default.parallelism")) {
+      rdd.context.defaultParallelism
+    } else {
+      rdds.map(_.partitions.length).max
+    }
+
+    // If the existing max partitioner is an eligible one, or its partitions number is larger
+    // than the default number of partitions, use the existing partitioner.
+    if (hasMaxPartitioner.nonEmpty && (isEligiblePartitioner(hasMaxPartitioner.get, rdds) ||
+        defaultNumPartitions < hasMaxPartitioner.get.getNumPartitions)) {
```

Comment:
This is the core change. I think it makes sense as it fixes a regression in https://github.com/apache/spark/pull/20002

If the partitioner is not eligible, but its numPartition is larger the the default one, we should still pick this partitioner instead of creating a new one.

Reply:
There are multiple cases here.

a) spark.default.parallelism is not set by user.
For this case, PR is a noop

b) maxPartitions is atleast an order higher than max partitioner

b.1) If spark.default.parallelism is not set, the PR is a noop.

b.2) spark.default.parallelism is explicitly set by user.

This is a change in behavior which has been introduced - rely on user specified value instead of trying to infer it when inferred value is off by atleast an order.


If users were setting suboptimal values for "spark.default.parallelism" - then there will be a change in behavior - though I would argue this is the *expected* behavior given documentation of 'spark.default.parallelism'

Reply:
It depends on how you define "default". In this case, if we can benefit from reusing an existing partitioner, we should pick that partitioner. If we want to respect `spark.default.parallelism` strictly, we should not reuse partitioner at all.

For this particular case, picking the existing partitioner is obviously a better choice and it was the behavior before #20002 , so I'm +1 on this change.

Reply:
> It depends on how you define "default".

I dont see an ambiguity here - am I missing something ?
To rephrase my point - this proposed PR has an impact only if user has explicitly set 'spark.default.parallelism' - else it is a noop.

What is the concern here ? Users have set incorrect values for spark.default.parallelism ?

> If we want to respect spark.default.parallelism strictly, we should not reuse partitioner at all.

I agree with you - we should not have - except that ship has sailed long long time back - since atleast 0.5 this has been the behavior in spark - I dont have context before that.
Historically, default parallelism was added later - using "largest partitioner if set or largest partition size when no partitioner is set" was the behavior. When default parallelism was introduced, probably (I guess) for backward compatible,  the behavior was continued.


#20002 surgically fixed only the case when inferred partition size was off by atleast an order.
When it is off by an order *and* if user has explicitly specified spark.default.parallelism, rely on user provided value - else preserve existing behavior.

Reply:
I think we all agree that reusing partitioner is an existing behavior and we should not stick to `spark.default.parallelism` here.

#20002 is good as it fixes a bad case where reusing partitioner slows down the query. And this PR surgically fixed one regression introduced by #20002 that, even if the existing partitioner is not eligible(has very little partitions), it's still better than fallback to default parallelism.